src/messages/MClientRequest.h

```diff
@@ -160,7 +161,28 @@ class MClientRequest : public Message {
 
   void decode_payload() {
     bufferlist::iterator p = payload.begin();
-    ::decode(head, p);
+
+    if (header.version >= 4) {
+      ::decode(head, p);
+    } else {
+      struct ceph_mds_request_head_legacy *old_mds_head =
+	(struct ceph_mds_request_head_legacy *)&(head.oldest_client_tid);
```

Comment:
Can't we just create a new (genuine) ceph_mds_request_head_legacy and then invoke a copy function to move it into our real one?
I realize just using the pointer is more efficient but hopefully this is an uncommon case, and doing it with pointer magic makes it really hard to find. Imagine in future when we do an incompat change and totally reorder all the member fields.